webui/react/src/components/UPlot/UPlotScatter/UPlotScatter.utils.ts

```diff
@@ -0,0 +1,218 @@
+
+export const X_INDEX = 0;
+export const Y_INDEX = 1;
+export const SIZE_INDEX = 2;
+export const FILL_INDEX = 3;
+export const STROKE_INDEX = 4;
+
+const DEFAULT_LOG_BASE = 10;
+const MIN_DIAMETER = 8;
+const MAX_DIAMETER = 60;
+const MIN_AREA = Math.PI * (MIN_DIAMETER / 2) ** 2;
+const MAX_AREA = Math.PI * (MAX_DIAMETER / 2) ** 2;
+const STROKE_WIDTH = 1;
+
+export type ColorFn = () => string;
+
+type BubbleFn<T = number | string> = (
+  value: number | null | undefined,
+  minValue: number,
+  maxValue: number
+) => T;
+
+export const getColorFn = (colorFn: unknown, fallbackColor: string): BubbleFn => {
+  const color = (colorFn as ColorFn)() || fallbackColor;
+  const colorFnRegex = new RegExp(/(rgba?\([^)]+\))\s+(rgba?\([^)]+\))/i);
+  const matches = color.match(colorFnRegex);
+  if (matches?.length !== 3) return () => color;
+
+  const minColor = matches[1];
+  const maxColor = matches[2];
+  const rgbaMin = str2rgba(minColor);
+  const rgbaMax = str2rgba(maxColor);
+  return (
+    value: number | null | undefined,
+    minValue: number,
+    maxValue: number,
+  ): string => {
+    if (value == null || minValue === maxValue || minValue == null) return minColor;
+    const percent = (value - minValue) / (maxValue - minValue);
+    const rgba = rgbaFromGradient(rgbaMin, rgbaMax, percent);
+    return rgba2str(rgba);
+  };
+};
+
+export const getMinMax = (u: uPlot, dataIndex: number): [ number, number ] => {
+  let minValue = Infinity;
+  let maxValue = -Infinity;
+
+  // Go through each series.
+  for (let i = 1; i < u.series.length; i++) {
+    const sizeData = u.data[i][dataIndex] as unknown as number[];
+
+    for (let j = 0; j < sizeData.length; j++) {
+      minValue = Math.min(minValue, sizeData[j]);
+      maxValue = Math.max(maxValue, sizeData[j]);
+    }
+  }
+
+  return [ minValue, maxValue ];
+};
+
+// quadratic scaling (px area)
+export const getSize = (
+  value: number | null | undefined,
+  minValue: number,
+  maxValue: number,
+): number => {
+  if (value == null || minValue === maxValue || minValue == null) return 0;
+  const percent = (value - minValue) / (maxValue - minValue);
+  const area = (MAX_AREA - MIN_AREA) * percent + MIN_AREA;
+  return Math.sqrt(area / Math.PI) * 2;
+};
+
+export const range = (
+  u: uPlot,
+  min: UPlotData,
+  max: UPlotData,
+  scaleKey: string,
+): Range<number> => {
+  // Return a standard range if there is not any valid data.
+  if (min == null || max == null) return [ 0, 100 ];
+
+  // When there is only one distinct value in the dataset.
+  if (min === max) {
+    const axisIndex = u.axes.findIndex(axis => axis.scale === scaleKey);
+    const axis = u.axes[axisIndex] as uPlot.Axis | undefined;
+    const getSplits = axis?.splits as UPlotAxisSplits | undefined;
+
+    if (/categorical/i.test(scaleKey)) {
+      // Using splits for categorical axis.
+      const splits = getSplits?.(u, axisIndex, min, max) || [];
+      if (splits.length > 1) return [ splits.first(), splits.last() ];
```

Comment:
hmmm good catch... this one is going to be a bit wonky because of uPlot's inability to handle a range that is limited to the 2 same values. And with categories, you can't really "pad the edges" with a small delta like the other scales. I ended up expanding it to `[ -1, 1 ]` as opposed to the default `[ 0 ]`. With categories the values are indices mapped to category values. So if there's only one unique value it would be one index value of `0`, hence `[ 0 ]`. I tried `[ -1, 1 ]` and the axis will at least render nicely but the point does not get plotted with the one value, which I think is ok for now.